datajoint/workflow-array-ephys
# Revise pytests/docker, markdown lint, pre-commit, spell check cleanup, NWB test update

I recommend reviewers select sets of commits to review, as some are formatting/spelling only, rather than substantive.

### Submission checklist

- [x] Reviewed all `Files changed` prior to selecting `Create Pull Request`.
- [x] Ran and passed all pytests, and ran the pipeline for features not covered by tests.
  - **NOTE** pytests have run piecemeal both locally and in docker. Will rerun each in full
- [X] Ran pre-commit checks, including black, flake8, etc.
- [X] Updated the changelog and if making a new release, updated the version.py.
- [ ] **N/A** Updated documentation, including concepts.md, tutorials.md, and relevant docstrings.
- [X] After merge, I will push a new tag to upstream (test* prefix if no new release).

### Submission details

In this pull request, I make the following changes:
- Add pre-commit checks, markdown lint, spell check config - for consistency across repos
   - `.pre-commit-config.yaml` - specify pre-commit checks
   - `cspell.json` - spell checker config
   - `CONTRIBUTING`, `CODE_OF_CONDUCT`, `README` - hard wrap, first line H1, line continuations on lists indented
   - `README` - add alt text for images
   - Many files - add proper line end
   - Many files - sort imports with isort
   - `setup.py` - add `noqa` flag for undeclared variable
- `CHANGELOG` - add brackets so older versions are recognized as links, add current changes
- `CONTRIBUTING` - update with new documentation
- `requirements_test` -> `requirements_dev` - more accurate regarding function. jupytext is dev work for notebooks, not tests
- Pytest adjustements
  - `pyproject.toml` - add notes for pytest flags and default settings, specify where tests are
  - `__init__.py` -> `conftest`
  - Add command line verbose and teardown options
  - Remove write_csv, combine csv writer fixtures to one
  - tests scripts - remove imports - no longer needed after shift to conftest
  - collapse `if verbose` to `with verbose_context` to centralize conditional
  - Add scope to fixtures to avoid premature teardown
  - `test_export.py` - shorten export frames to 250
  - As nwb tests exist here, needed to re-add lab imports with corresponding CSVs - including support to add to new `project` schema in Element Lab
- Docker
  - Add `IS_DOCKER` env variable to allow tests to pass locally
  - `setup.sh` allows env variables to dictate if tests are run when setup, collapsing need for `dev` vs `test` dockerfiles
- Documentation: `analysis.py` needed docstring work for API doc rendering.  

[pytest_coverage.log](https://github.com/datajoint/workflow-array-ephys/files/10449357/pytest_coverage.log)
